io.fits errors with tables containing strings

The following code contains a set of unit tests for creating FITS binary tables with strings.  They pass under python 2.7 but three of the five fail under python 3.5.  This is also available as a [public gist](https://gist.github.com/sbailey/f5971fc3d28952a39a4f9eff7d2c2408) but I don't promise to keep that around unchanged forever, so the code is also copy and pasted below.

The tests that fail are:
- writing a numpy structured array with a column of strings using `astropy.io.fits.writeto` and reading back in with `astropy.io.fits.getdata`: the data are munged.  This works when writing the same data converted to a Table object, and also works when creating the BinTableHDU from columns by hand.  This test is a demonstration of the failure reported in #5268 .
- writing then reading a numpy structured array containing a 2D column of strings using `astropy.io.fits.writeto` and reading back in with `astropy.io.fits.getdata`: the dtype seems to have mixed up the number of unicode bytes vs. the number of characters to write.
- writing then reading an astropy.table.Table containing a 2D column of strings: same problem with the dtype

``` python
import unittest
from uuid import uuid4
import os

import numpy as np
from astropy.io import fits
from astropy.table import Table, Column

class _TestTable(unittest.TestCase):
    '''
    Base class for testing FITS tables.

    Subclass this and add setupClass to define attributes to test:
        data (numpy structured array)
        table (astropy Table)
    '''

    def setUp(self):
        self.testfile = 'test-{}.fits'.format(uuid4())

    def tearDown(self):
        if os.path.exists(self.testfile):
            os.remove(self.testfile)

    def test_fits(self):
        fits.writeto(self.testfile, self.data)
        dx = fits.getdata(self.testfile)
        self.assertEqual(self.data['x'].dtype, dx['x'].dtype)
        self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
        self.assertTrue(np.all(self.data['x'] == dx['x']), 'x: {} != {}'.format(self.data['x'], dx['x']))
        self.assertTrue(np.all(self.data['y'] == dx['y']), 'y: {} != {}'.format(self.data['y'], dx['y']))

    def test_table(self):
        self.table.write(self.testfile)
        tx = Table.read(self.testfile)
        self.assertEqual(self.table['x'].dtype, tx['x'].dtype)
        self.assertEqual(self.table['y'].dtype, tx['y'].dtype)
        self.assertTrue(np.all(self.table['x'] == tx['x']), 'x: {} != {}'.format(self.table['x'], tx['x']))
        self.assertTrue(np.all(self.table['y'] == tx['y']), 'y: {} != {}'.format(self.table['y'], tx['y']))

class TestTable1DColumn(_TestTable):
    @classmethod
    def setUpClass(cls):
        dtype = [
            ('x', (str, 5)),    #- 1D column of 5-character strings
            ('y', (str, 3)),    #- !D column of 3-character strings
        ]
        data = np.zeros(2, dtype=dtype)
        data['x'] = ['abcde', 'xyz']
        data['y'] = ['A', 'BCX']
        table = Table(data)
        cls.data = data
        cls.table = table

    def test_manual_fits(self):
        '''Construct the BinTableHDU by hand'''
        col1 = fits.Column(name='x', format='5A', array=self.data['x'])
        col2 = fits.Column(name='y', format='3A', array=self.data['y'])
        cols = fits.ColDefs([col1, col2])
        tbhdu = fits.BinTableHDU.from_columns(cols)
        tbhdu.writeto(self.testfile)
        dx = fits.getdata(self.testfile)
        self.assertEqual(self.data['x'].dtype, dx['x'].dtype)
        self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
        self.assertTrue(np.all(self.data['x'] == dx['x']), 'x: {} != {}'.format(self.data['x'], dx['x']))
        self.assertTrue(np.all(self.data['y'] == dx['y']), 'y: {} != {}'.format(self.data['y'], dx['y']))


class TestTable2DColumn(_TestTable):
    @classmethod
    def setUpClass(cls):
        dtype = [
            ('x', (str, 5)),        #- 1D column of 5-character strings
            ('y', (str, 3), (4,)),  #- 2D column; each row is four 3-char strings
        ]
        data = np.zeros(2, dtype=dtype)
        data['x'] = ['abcde', 'xyz']
        data['y'][0] = ['A', 'BC', 'DEF', '123']
        data['y'][1] = ['X', 'YZ', 'PQR', '999']
        table = Table(data)
        cls.data = data
        cls.table = table

    #- I'm not sure how to create 2D column of strings by hand analogous to
    #- the 1D test case; see https://github.com/astropy/astropy/issues/5279
    # def test_manual_fits(self):
    #     '''Construct the BinTableHDU by hand
    #     
    #     When these data are successfully written under py2.7, they have keywords
    #     TTYPE1  = 'x       '
    #     TFORM1  = '5A      '
    #     TTYPE2  = 'y       '
    #     TFORM2  = '12A     '
    #     TDIM2   = '(3,4)   '
    #     
    #     Trying to replicate that logic here with dim=(3,4) doesn't work
    #     '''
    #     col1 = fits.Column(name='x', format='5A', array=self.data['x'])
    #     col2 = fits.Column(name='y', format='12A', array=self.data['y'], dim=(3,4))
    #     cols = fits.ColDefs([col1, col2])
    #     tbhdu = fits.BinTableHDU.from_columns(cols)
    #     tbhdu.writeto(self.testfile)
    #     dx = fits.getdata(self.testfile)
    #     self.assertEqual(self.data['x'].dtype, dx['x'].dtype)
    #     self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
    #     self.assertTrue(np.all(self.data['x'] == dx['x']), 'x: {} != {}'.format(self.data['x'], dx['x']))
    #     self.assertTrue(np.all(self.data['y'] == dx['y']), 'y: {} != {}'.format(self.data['y'], dx['y']))


if __name__ == '__main__':
    ### unittest.main()
    t1 = unittest.defaultTestLoader.loadTestsFromTestCase(TestTable1DColumn)
    t2 = unittest.defaultTestLoader.loadTestsFromTestCase(TestTable2DColumn)
    unittest.TextTestRunner(verbosity=2).run(unittest.TestSuite([t1,t2]))
```

Running these pass under python 2.7 but fail under python 3.5 with:

``` python
(py3) py3 $ python test-fits-table.py 
test_fits (__main__.TestTable1DColumn) ... WARNING: File may have been truncated: actual file length (8636) is smaller than the expected size (8640) [astropy.io.fits.file]
FAIL
test_manual_fits (__main__.TestTable1DColumn)
Construct the BinTableHDU by hand ... ok
test_table (__main__.TestTable1DColumn) ... ok
test_fits (__main__.TestTable2DColumn) ... WARNING: File may have been truncated: actual file length (8566) is smaller than the expected size (8640) [astropy.io.fits.file]
FAIL
test_table (__main__.TestTable2DColumn) ... FAIL

======================================================================
FAIL: test_fits (__main__.TestTable1DColumn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test-fits-table.py", line 30, in test_fits
    self.assertTrue(np.all(self.data['x'] == dx['x']), 'x: {} != {}'.format(self.data['x'], dx['x']))
AssertionError: False is not true : x: ['abcde' 'xyz'] != ['abcde' 'zBCX']

======================================================================
FAIL: test_fits (__main__.TestTable2DColumn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test-fits-table.py", line 29, in test_fits
    self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
AssertionError: dtype('<U3') != dtype('<U12')

======================================================================
FAIL: test_table (__main__.TestTable2DColumn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test-fits-table.py", line 37, in test_table
    self.assertEqual(self.table['y'].dtype, tx['y'].dtype)
AssertionError: dtype('<U3') != dtype('<U12')

----------------------------------------------------------------------
Ran 5 tests in 0.100s

FAILED (failures=3)
```

EDIT: Added syntax highlighting.

I have run the test of this issue against the current version on main.
The exact same error is thrown.
<details>
<summary>Test script</summary>

```py
#- unit tests of astropy.io.fits
#- Stephen Bailey, Lawrence Berkeley Lab

from __future__ import unicode_literals

import unittest
from uuid import uuid4
import os

import numpy as np
from astropy.io import fits
from astropy.table import Table, Column

class _TestTable(unittest.TestCase):
    '''
    Base class for testing FITS tables.
    Subclass this and add setupClass to define attributes to test:
        data (numpy structured array)
        table (astropy Table)
    '''

    def setUp(self):
        self.testfile = 'test-{}.fits'.format(uuid4())

    def tearDown(self):
        if os.path.exists(self.testfile):
            os.remove(self.testfile)

    def test_fits_writeto(self):
        '''test convenience functions io.fits.writeto / getdata'''
        fits.writeto(self.testfile, self.data)
        dx = fits.getdata(self.testfile)
        self.assertEqual(self.data['x'].dtype, dx['x'].dtype)
        self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
        self.assertTrue(np.all(self.data['x'] == dx['x']), 'x: {} != {}'.format(self.data['x'], dx['x']))
        self.assertTrue(np.all(self.data['y'] == dx['y']), 'y: {} != {}'.format(self.data['y'], dx['y']))
    
    def test_fits_bintable(self):
        '''Test fits.BinTableHDU(data) and avoid convenience functions'''
        hdu0 = fits.PrimaryHDU()
        hdu1 = fits.BinTableHDU(self.data)
        hx = fits.HDUList([hdu0, hdu1])
        hx.writeto(self.testfile)
        # dx = fits.getdata(self.testfile)
        fx = fits.open(self.testfile)
        dx = fx[1].data
        fx.close()
        self.assertEqual(self.data['x'].dtype, dx['x'].dtype)
        self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
        self.assertTrue(np.all(self.data['x'] == dx['x']), 'x: {} != {}'.format(self.data['x'], dx['x']))
        self.assertTrue(np.all(self.data['y'] == dx['y']), 'y: {} != {}'.format(self.data['y'], dx['y']))

    def test_table(self):
        self.table.write(self.testfile)
        tx = Table.read(self.testfile)
        self.assertEqual(self.table['x'].dtype, tx['x'].dtype)
        self.assertEqual(self.table['y'].dtype, tx['y'].dtype)
        self.assertTrue(np.all(self.table['x'] == tx['x']), 'x: {} != {}'.format(self.table['x'], tx['x']))
        self.assertTrue(np.all(self.table['y'] == tx['y']), 'y: {} != {}'.format(self.table['y'], tx['y']))

class TestTable1DColumn(_TestTable):
    @classmethod
    def setUpClass(cls):
        dtype = [
            (str('x'), (str, 5)),    #- 1D column of 5-character unicode strings (Python3)
            (str('y'), (str, 3)),    #- !D column of 3-character unicode strings (Python3)
        ]
        data = np.zeros(2, dtype=dtype)
        data['x'] = ['abcde', 'xyz']
        data['y'] = ['A', 'BCX']
        table = Table(data)
        cls.data = data
        cls.table = table

    def test_fits_columns(self):
        '''Test BinTableHDU.from_columns()'''
        col1 = fits.Column(name=str('x'), format='5A', array=self.data['x'])
        col2 = fits.Column(name=str('y'), format='3A', array=self.data['y'])
        cols = fits.ColDefs([col1, col2])
        tbhdu = fits.BinTableHDU.from_columns(cols)
        tbhdu.writeto(self.testfile)
        dx = fits.getdata(self.testfile)
        self.assertEqual(self.data['x'].dtype, dx['x'].dtype)
        self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
        self.assertTrue(np.all(self.data['x'] == dx['x']), 'x: {} != {}'.format(self.data['x'], dx['x']))
        self.assertTrue(np.all(self.data['y'] == dx['y']), 'y: {} != {}'.format(self.data['y'], dx['y']))
        

class TestTable2DColumn(_TestTable):
    @classmethod
    def setUpClass(cls):
        dtype = [
            (str('x'), (str, 5)),        #- 1D column of 5-character unicode strings
            (str('y'), (str, 3), (4,)),  #- 2D column; each row is four 3-char unicode strings
        ]
        data = np.zeros(2, dtype=dtype)
        data['x'] = ['abcde', 'xyz']
        data['y'][0] = ['A', 'BC', 'DEF', '123']
        data['y'][1] = ['X', 'YZ', 'PQR', '999']
        table = Table(data)
        cls.data = data
        cls.table = table

    def test_fits_columns(self):
        '''Test BinTableHDU.from_columns()'''
        col1 = fits.Column(name=str('x'), format='5A', array=self.data['x'])
        col2 = fits.Column(name=str('y'), format='12A', array=self.data['y'], dim='(3,4)')
        cols = fits.ColDefs([col1, col2])
        tbhdu = fits.BinTableHDU.from_columns(cols)
        tbhdu.writeto(self.testfile)
        dx = fits.getdata(self.testfile)
        self.assertEqual(self.data['x'].dtype, dx['x'].dtype)
        self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
        self.assertTrue(np.all(self.data['x'] == dx['x']), 'x: {} != {}'.format(self.data['x'], dx['x']))
        self.assertTrue(np.all(self.data['y'] == dx['y']), 'y: {} != {}'.format(self.data['y'], dx['y']))

                
if __name__ == '__main__':
    ### unittest.main()
    t1 = unittest.defaultTestLoader.loadTestsFromTestCase(TestTable1DColumn)
    t2 = unittest.defaultTestLoader.loadTestsFromTestCase(TestTable2DColumn)
    unittest.TextTestRunner(verbosity=2).run(unittest.TestSuite([t1, t2]))
```
</details>

<details>
<summary>Here is the traceback:</summary>

```pytb
(astropydevenv) [cmarmo@2603-800c-2300-a5d0-d22b-4854-594d-178b tests]$ python 5280-astropy.py 
test_fits_bintable (__main__.TestTable1DColumn)
Test fits.BinTableHDU(data) and avoid convenience functions ... ok
test_fits_columns (__main__.TestTable1DColumn)
Test BinTableHDU.from_columns() ... ok
test_fits_writeto (__main__.TestTable1DColumn)
test convenience functions io.fits.writeto / getdata ... ok
test_table (__main__.TestTable1DColumn) ... FAIL
test_fits_bintable (__main__.TestTable2DColumn)
Test fits.BinTableHDU(data) and avoid convenience functions ... FAIL
test_fits_columns (__main__.TestTable2DColumn)
Test BinTableHDU.from_columns() ... ok
test_fits_writeto (__main__.TestTable2DColumn)
test convenience functions io.fits.writeto / getdata ... FAIL
test_table (__main__.TestTable2DColumn) ... FAIL

======================================================================
FAIL: test_table (__main__.TestTable1DColumn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/cmarmo/software/tests/5280-astropy.py", line 56, in test_table
    self.assertEqual(self.table['x'].dtype, tx['x'].dtype)
AssertionError: dtype('<U5') != dtype('S5')

======================================================================
FAIL: test_fits_bintable (__main__.TestTable2DColumn)
Test fits.BinTableHDU(data) and avoid convenience functions
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/cmarmo/software/tests/5280-astropy.py", line 49, in test_fits_bintable
    self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
AssertionError: dtype('<U3') != dtype('<U12')

======================================================================
FAIL: test_fits_writeto (__main__.TestTable2DColumn)
test convenience functions io.fits.writeto / getdata
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/cmarmo/software/tests/5280-astropy.py", line 34, in test_fits_writeto
    self.assertEqual(self.data['y'].dtype, dx['y'].dtype)
AssertionError: dtype('<U3') != dtype('<U12')

======================================================================
FAIL: test_table (__main__.TestTable2DColumn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/cmarmo/software/tests/5280-astropy.py", line 56, in test_table
    self.assertEqual(self.table['x'].dtype, tx['x'].dtype)
AssertionError: dtype('<U5') != dtype('S5')

----------------------------------------------------------------------
Ran 8 tests in 0.067s

FAILED (failures=4)
```
</details>

I could fix the `AssertionError: dtype('<U5') != dtype('S5')` modifying `test_table` as following
```py
    def test_table(self):
        self.table.write(self.testfile)
        tx = Table.read(self.testfile, character_as_bytes=False)
        self.assertEqual(self.table['x'].dtype, tx['x'].dtype)
        self.assertEqual(self.table['y'].dtype, tx['y'].dtype)
        self.assertTrue(np.all(self.table['x'] == tx['x']), 'x: {} != {}'.format(self.table['x'], tx['x']))
        self.assertTrue(np.all(self.table['y'] == tx['y']), 'y: {} != {}'.format(self.table['y'], tx['y']))
```
If I understand correctly, this is due to the fact that string dtypes are written in unicode by numpy, but the astropy `read_table_fits` in [connect.py](https://github.com/astropy/astropy/blob/ef6c4b15537523a1dedcbf984350a67d94cb7272/astropy/io/fits/connect.py#L114-L115) reads by default in binary. Adding `character_as_bytes=False` will read in unicode.
This is not a bug, perhaps this could be made more explicit in the documentation.

I was unable to fix the `AssertionError: dtype('<U3') != dtype('<U12')` but I tracked it to [table.py](https://github.com/astropy/astropy/blob/ef6c4b15537523a1dedcbf984350a67d94cb7272/astropy/io/fits/hdu/table.py#L123).
This is where columns move from the format `[('x', 'S5'), ('y', 'S3', (4,))]` to
```
ColDefs(
    name = 'x'; format = '5A'
    name = 'y'; format = '48A'; dim = '(12,4)'
)
```
I guess, there is some automatic mechanism to convert to ColDefs that I couldn't understand.

Hope this can help in moving forward.

--- a/astropy/io/fits/column.py
+++ b/astropy/io/fits/column.py
@@ -1437,10 +1437,10 @@ class ColDefs(NotifierMixin):
             format = self._col_format_cls.from_recformat(ftype)
 
             # Determine the appropriate dimensions for items in the column
-            # (typically just 1D)
             dim = array.dtype[idx].shape[::-1]
             if dim and (len(dim) > 0 or 'A' in format):
                 if 'A' in format:
+                    # should take into account multidimensional items in the column
                     dimel = int(re.findall('[0-9]+', str(ftype.subdtype[0]))[0])
                     # n x m string arrays must include the max string
                     # length in their dimensions (e.g. l x n x m)
@@ -2370,7 +2370,7 @@ def _convert_record2fits(format):
                              and dtype.subdtype[0].char == 'U'):
         # Unicode dtype--itemsize is 4 times actual ASCII character length,
         # which what matters for FITS column formats
-        # Use dtype.base and dtype.subdtype --dtype may be a multi-dimensional dtype
+        # Use dtype.base and dtype.subdtype --dtype for multi-dimensional items
         itemsize = itemsize // 4
 
     option = str(itemsize)

--- a/astropy/io/fits/column.py
+++ b/astropy/io/fits/column.py
@@ -1441,9 +1441,10 @@ class ColDefs(NotifierMixin):
             dim = array.dtype[idx].shape[::-1]
             if dim and (len(dim) > 0 or 'A' in format):
                 if 'A' in format:
+                    dimel = int(re.findall('[0-9]+', str(ftype.subdtype[0]))[0])
                     # n x m string arrays must include the max string
                     # length in their dimensions (e.g. l x n x m)
-                    dim = (array.dtype[idx].base.itemsize,) + dim
+                    dim = (dimel,) + dim
                 dim = '(' + ','.join(str(d) for d in dim) + ')'
             else:
                 dim = None
@@ -2365,10 +2366,11 @@ def _convert_record2fits(format):
     recformat, kind, dtype = _dtype_to_recformat(format)
     shape = dtype.shape
     itemsize = dtype.base.itemsize
-    if dtype.char == 'U':
+    if dtype.char == 'U' or (dtype.subdtype is not None
+                             and dtype.subdtype[0].char == 'U'):
         # Unicode dtype--itemsize is 4 times actual ASCII character length,
         # which what matters for FITS column formats
-        # Use dtype.base--dtype may be a multi-dimensional dtype
+        # Use dtype.base and dtype.subdtype --dtype may be a multi-dimensional dtype
         itemsize = itemsize // 4
 
     option = str(itemsize)

--- a/astropy/io/fits/tests/test_table.py
+++ b/astropy/io/fits/tests/test_table.py
@@ -969,6 +969,54 @@ class TestTableFunctions(FitsTestCase):
             assert header['TNULL2'] == 'b'
             assert header['TNULL3'] == 2.3
 
+    def test_multidimension_table_from_numpy_rec_columns(self):
+        """Regression test for https://github.com/astropy/astropy/issues/5280
+        and https://github.com/astropy/astropy/issues/5287
+
+        multidimentional tables can now be written with the correct TDIM.
+        Author: Stephen Bailey.
+        """
+
+        dtype = [
+            (str('x'), (str, 5)),        # 1D column of 5-character strings
+            (str('y'), (str, 3), (4,)),  # 2D column; each row is four 3-char strings
+        ]
+        data = np.zeros(2, dtype=dtype)
+        data['x'] = ['abcde', 'xyz']
+        data['y'][0] = ['A', 'BC', 'DEF', '123']
+        data['y'][1] = ['X', 'YZ', 'PQR', '999']
+        table = Table(data)
+
+        testfile = self.temp('test.fits')
+        # Test convenience functions io.fits.writeto / getdata
+        fits.writeto(testfile, data)
+        dx = fits.getdata(testfile)
+        assert data['x'].dtype == dx['x'].dtype
+        assert data['y'].dtype == dx['y'].dtype
+        assert np.all(data['x'] == dx['x']), 'x: {} != {}'.format(data['x'], dx['x'])
+        assert np.all(data['y'] == dx['y']), 'y: {} != {}'.format(data['y'], dx['y'])
+
+        # Test fits.BinTableHDU(data) and avoid convenience functions
+        hdu0 = fits.PrimaryHDU()
+        hdu1 = fits.BinTableHDU(data)
+        hx = fits.HDUList([hdu0, hdu1])
+        hx.writeto(testfile, overwrite=True)
+        fx = fits.open(testfile)
+        dx = fx[1].data
+        fx.close()
+        assert data['x'].dtype == dx['x'].dtype
+        assert data['y'].dtype == dx['y'].dtype
+        assert np.all(data['x'] == dx['x']), 'x: {} != {}'.format(data['x'], dx['x'])
+        assert np.all(data['y'] == dx['y']), 'y: {} != {}'.format(data['y'], dx['y'])
+
+        # Test Table write and read
+        table.write(testfile, overwrite=True)
+        tx = Table.read(testfile, character_as_bytes=False)
+        assert table['x'].dtype == tx['x'].dtype
+        assert table['y'].dtype == tx['y'].dtype
+        assert np.all(table['x'] == tx['x']), 'x: {} != {}'.format(table['x'], tx['x'])
+        assert np.all(table['y'] == tx['y']), 'y: {} != {}'.format(table['y'], tx['y'])
+
     def test_mask_array(self):
         t = fits.open(self.data('table.fits'))
         tbdata = t[1].data

--- a/astropy/io/fits/tests/test_table.py
+++ b/astropy/io/fits/tests/test_table.py
@@ -987,22 +987,20 @@ class TestTableFunctions(FitsTestCase):
         data['y'][1] = ['X', 'YZ', 'PQR', '999']
         table = Table(data)
 
-        testfile = self.temp('test.fits')
         # Test convenience functions io.fits.writeto / getdata
-        fits.writeto(testfile, data)
-        with fits.open(testfile):
-            dx = fits.getdata(testfile)
-            assert data['x'].dtype == dx['x'].dtype
-            assert data['y'].dtype == dx['y'].dtype
-            assert np.all(data['x'] == dx['x']), 'x: {} != {}'.format(data['x'], dx['x'])
-            assert np.all(data['y'] == dx['y']), 'y: {} != {}'.format(data['y'], dx['y'])
+        fits.writeto(self.temp('test.fits'), data)
+        dx = fits.getdata(self.temp('test.fits'))
+        assert data['x'].dtype == dx['x'].dtype
+        assert data['y'].dtype == dx['y'].dtype
+        assert np.all(data['x'] == dx['x']), 'x: {} != {}'.format(data['x'], dx['x'])
+        assert np.all(data['y'] == dx['y']), 'y: {} != {}'.format(data['y'], dx['y'])
 
         # Test fits.BinTableHDU(data) and avoid convenience functions
         hdu0 = fits.PrimaryHDU()
         hdu1 = fits.BinTableHDU(data)
         hx = fits.HDUList([hdu0, hdu1])
-        hx.writeto(testfile, overwrite=True)
-        fx = fits.open(testfile)
+        hx.writeto(self.temp('test2.fits'))
+        fx = fits.open(self.temp('test2.fits'))
         dx = fx[1].data
         fx.close()
         assert data['x'].dtype == dx['x'].dtype
@@ -1011,8 +1009,8 @@ class TestTableFunctions(FitsTestCase):
         assert np.all(data['y'] == dx['y']), 'y: {} != {}'.format(data['y'], dx['y'])
 
         # Test Table write and read
-        table.write(testfile, overwrite=True)
-        tx = Table.read(testfile, character_as_bytes=False)
+        table.write(self.temp('test3.fits'))
+        tx = Table.read(self.temp('test3.fits'), character_as_bytes=False)
         assert table['x'].dtype == tx['x'].dtype
         assert table['y'].dtype == tx['y'].dtype
         assert np.all(table['x'] == tx['x']), 'x: {} != {}'.format(table['x'], tx['x'])

--- a/astropy/io/fits/tests/test_table.py
+++ b/astropy/io/fits/tests/test_table.py
@@ -990,11 +990,12 @@ class TestTableFunctions(FitsTestCase):
         testfile = self.temp('test.fits')
         # Test convenience functions io.fits.writeto / getdata
         fits.writeto(testfile, data)
-        dx = fits.getdata(testfile)
-        assert data['x'].dtype == dx['x'].dtype
-        assert data['y'].dtype == dx['y'].dtype
-        assert np.all(data['x'] == dx['x']), 'x: {} != {}'.format(data['x'], dx['x'])
-        assert np.all(data['y'] == dx['y']), 'y: {} != {}'.format(data['y'], dx['y'])
+        with fits.open(testfile):
+            dx = fits.getdata(testfile)
+            assert data['x'].dtype == dx['x'].dtype
+            assert data['y'].dtype == dx['y'].dtype
+            assert np.all(data['x'] == dx['x']), 'x: {} != {}'.format(data['x'], dx['x'])
+            assert np.all(data['y'] == dx['y']), 'y: {} != {}'.format(data['y'], dx['y'])
 
         # Test fits.BinTableHDU(data) and avoid convenience functions
         hdu0 = fits.PrimaryHDU()

--- a/docs/changes/io.fits/12863.bugfix.rst
+++ b/docs/changes/io.fits/12863.bugfix.rst
@@ -1,0 +1,2 @@
+Fix write and read FITS tables with multidimensional items, using ``from_columns``
+without previousely defined ``ColDefs`` structure.

